sw/device/lib/dif/dif_kmac.c

```diff
@@ -4,4 +4,478 @@
+    return kDifKmacError;
+  }
+
+  // Copy the message one byte at a time.
+  // This could be sped up copying a word at a time but be careful
+  // about message endianness (e.g. only copy a word at a time when in
+  // little-endian mode).
+  for (size_t i = 0; i < len; ++i) {
+    mmio_region_write8(kmac->params.base_addr, KMAC_MSG_FIFO_REG_OFFSET,
+                       ((const uint8_t *)msg)[i]);
+  }
+
+  if (processed != NULL) {
+    *processed = len;
+  }
+  return kDifKmacOk;
+}
+
+dif_kmac_result_t dif_kmac_squeeze(dif_kmac_t *kmac, void *out, size_t len,
+                                   size_t *processed) {
+  if (kmac == NULL || (out == NULL && len != 0)) {
+    return kDifKmacBadArg;
+  }
+
+  // Set `processed` to 0 so we can return early without setting it again.
+  if (processed != NULL) {
+    *processed = 0;
+  }
+
+  // Move into squeezing state (if not already in it).
+  // Do this even if the length requested is 0 or too big.
+  if (!kmac->squeezing) {
+    kmac->squeezing = true;
+
+    // Issue squeeze command.
+    uint32_t cmd_reg = bitfield_field32_write(0, KMAC_CMD_CMD_FIELD,
+                                              KMAC_CMD_CMD_VALUE_PROCESS);
+    mmio_region_write32(kmac->params.base_addr, KMAC_CMD_REG_OFFSET, cmd_reg);
+  }
+
+  // If the operation has a fixed length output then the total number of bytes
+  // requested must not exceed that length.
+  if (kmac->d != 0 && len > (kmac->d - kmac->offset)) {
+    return kDifKmacError;
+  }
+
+  if (len == 0) {
+    return kDifKmacOk;
+  }
+
+  // Poll the status register until in the 'squeeze' state.
+  while (true) {
+    if (is_state_squeeze(kmac->params)) {
+      break;
+    }
+    // TODO(#6248): check for error.
+  }
+
+  // Read out the state shares.
+  while (len > 0) {
+    size_t n = len;
+    size_t remaining = (kmac->d == 0 ? kmac->r : kmac->d) - kmac->offset;
+    if (n > remaining) {
+      n = remaining;
+    }
+    if (n == 0) {
+      // TODO(XOF): request more state.
+      return kDifKmacError;
+    }
+
+    // The output state register is divided into two shares. A word-aligned
+    // region containing the requested bytes is copied from each share and
+    // combined using XOR. The requested bytes are then copied from the combined
+    // state into the byte-aligned destination buffer.
+    //
+    //        kmac->offset
+    //                   |
+    //   word boundary   | kmac->offset + n   word boundary
+    //               |   |                |   |
+    // +=============+===+================+===+=============+
+    // | state share |   |                |   |             |
+    // +=============+===+================+===+=============+
+    //               |                        |
+    //               | <-- region to copy --> |
+    //    state_offset                        |
+    //                 state_offset + state_len
+
+    // Mask 2 lowest bits.
+    static const uint32_t kMask = 0x3u;
```

Comment:
> 1. Is KMAC_STATE_REG_OFFSET + 0x0 the digest_share_0 offset and + 0x100 digest_share_1 offset?

Basically, yes. Both shares are in one `STATE` register but you could think of it as `STATE_SHARE_0` at 0x400 and `STATE_SHARE_1` at 0x500.

> 2. Why not read the entire digest in uint32_t words? I was poking around the DV test code and couldn't figure out why these offset calculations are needed.

The current API allows the user to call `dif_kmac_squeeze` multiple times with arbitrary output lengths during a single hash operation. We need to track an offset into the output data to enable this. This approach gives maximum flexibility to the caller at the expense of implementation complexity. This is a useful pattern to use when the output data might need to be split into multiple fixed size packets for example. We do need to allow multiple calls to `dif_kmac_squeeze` if we want to allow non-blocking XOF operations but we don't necessarily need to allow arbitrary offsets for this.

Making the out-param a `uint32_t *` rather than a `void *` would simplify the function and I'm happy to go down that route if we aren't interested in allowing an arbitrary number of bytes to be written out.

> For the XFO modes, if digest_size > r, I'd imagine we have to run another cycle and then read again the state (digest) register starting at offset 0, is that correct?

Yes, that is correct. In blocking mode (`processed == NULL`) we'd trigger the computation and wait while in non-blocking mode we'd trigger the computation, write the number of bytes read to `processed` and then return.

Reply:
Given that these outputs will be consumed explicitly as digests, I think it is safe to use an `uint32_t *` out-param. So far all use cases and planned tests are expected to treat the output as a digest.

Reply:
I a `void *` use case arises, then we can add additional complexity, but at this moment I prefer to go with the simpler implementation.

Reply:
Done. The `dif_kmac_squeeze` out-param is now a `uint32_t *`.